contents/docs/contributing.md

```diff
@@ -29,9 +29,22 @@ If you're planning to work on a bigger feature, and it's not in the list of issu
 
 ## Testing
 
+### Backend
 We expect all backend code to have corresponding tests. You can run the tests by running `bin/tests`. To specify a specific test (rather than running all tests) you can run `bin/tests posthog.api.test.test_action` for example.
 
-At the moment, we do not have many frontend tests. Do make sure that each page you've touched (or that uses a component you touched) still functions as expected.
+### Frontend
+We use [cypress](https://www.cypress.io/) for functional frontend tests. Cypress is used to mock user interface interactions on a demo instance. We've currently targeted test coverage for the most used areas of posthog and are working to increase coverage. 
+
+We follow the given, when, then framework. Each test should provide a given (initial setup for the flow or feature being tested), when (an action performed), and then (an assertion that verifies resulting behavior). To keep our components flexible and easily maintainable within tests, we use a `dataattr` pattern where we add a page-unique identifier prop to react components for testing purposes. Then, to identify an element in a cypress tests we use `cy.get('[dataattr=some-identifier]')...`. 
```

Comment:
Hi, I'm wondering about this. Normally the `data-*` namespace in HTML attributes is reserved for whatever custom data we want to store on DOM nodes. Other attributes are reserved and/or whitelisted according to whichever HTML specification. As in AFAIK `<div data-attr="bla">hello</div>` is valid DOM, but `<div randomstring="bla">hello</div>` and `<div dataattr="bla">hello</div>` are not. 

I could be wrong though... but I think it's worth checking.

On a related note, could it be worth "officializing" the data-attr pattern somehow in posthog? Since we are adding all this extra instrumentation into posthog app, couldn't we also use this when capturing events on the app? And couldn't other people using posthog-js also use this pattern to have better intelligence into their actions?

Reply:
Do you mean not valid as it would error out? In testing, the only warning that React gives is if you use a camelcase attribute that's unrecognized (which is why i opted for dataattr rather than dataAttr). The warning suggests that you use all lowercase for custom attributes that you want to appear in the DOM. The docs [here](https://reactjs.org/blog/2017/09/08/dom-attributes-in-react-16.html) explain this behavior. 
<img width="507" alt="Screen Shot 2020-05-18 at 4 30 27 PM" src="https://user-images.githubusercontent.com/13127476/82256760-ea7aa500-9924-11ea-9b78-f45ed8d63c57.png">

RE: second point. I guess we could—what exactly were you imagining the tags would do analytics wise?

Reply:
Ah, what about just `data-attr` then, with a dash?

![image](https://user-images.githubusercontent.com/53387/82258264-d0a78500-9959-11ea-90b9-2b1676934d91.png)

React seems to allow passing it like this:
https://reactjs.org/blog/2017/09/08/dom-attributes-in-react-16.html#data-and-aria-attributes

Having it part of the [dataset](https://developer.mozilla.org/en-US/docs/Web/API/HTMLOrForeignElement/dataset) instead of a made up new attribute has some benefits. For one it'll never clash with some real standard attribute W3C might come up with, though the chance for `dataattr` is quite small. For another, since it's part of the standard, who knows what (de)serialization issues can be handled for us

[The docs](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) do say "`data-*` attributes allow us to store extra information on standard, semantic HTML elements without other hacks such as non-standard attributes, extra properties on DOM, or Node.setUserData().", which sounds like what we're doing.

---

For the second point, I'm not sure. Just throwing it out there. The simplest is that in the events table, instead of "clicked `<input>`" the first column can just say "clicked <input data-attr="something">`" to add a bit more context. Kind of like we should do with aria-labels (issue #791)

Reply:
Ah I see, good call. Will make the switch. The dash also makes it more readable